hera_cal/lst_stack/io.py

```diff
@@ -270,18 +267,63 @@ def create_lstbin_output_file(
+    flags: np.ndarray,
+    nsamples: np.ndarray,
+    time_index: int | None = None,
 ):
-    """Write a baseline slice to a file."""
-    with h5py.File(fl, "r+") as f:
-        ntimes = int(f["Header"]["Ntimes"][()])
-        timefirst = bool(f["Header"]["time_axis_faster_than_bls"][()])
-        if not timefirst and ntimes > 1:
-            raise NotImplementedError("Can only do time-first files for now.")
-
-        slc = slice(slc.start * ntimes, slc.stop * ntimes, 1)
-        f["Data"]["visdata"][slc] = data.reshape((-1, data.shape[2], data.shape[3]))
-        f["Data"]["flags"][slc] = flags.reshape((-1, data.shape[2], data.shape[3]))
-        f["Data"]["nsamples"][slc] = nsamples.reshape(
-            (-1, data.shape[2], data.shape[3])
+    """Write a baseline slice to a file.
+
+    This is a specialized function that writes a slice of baselines to a pre-existing
+    uvh5 file. The file must be in a specific format for this function to work, namely
+    it must be a .uvh5 file with a rectangular blt axis, where the time axis is the
+    outer most axis. Such a file is created by the :func:`create_lstbin_output_file`
+    function.
+
+    This function is useful when lst-averaging, to write out partial results to disk,
+    when looping over baselines and/or lsts.
+
+    Parameters
+    ----------
+    fl : Path or FastUVH5Meta
+        Path to the file or an open FastUVH5Meta object.
+    baseline_slice : slice
+        Slice of baselines to write to the file -- these must be the baseline indices
+        (NOT blt indices) *in the file* to which the input data corresponds.
+    data : np.ndarray
+        Data to write to the file. Must have shape ``(nbl_slice, nfreqs, npols)`` if
+        ``time_index`` is not None, or ``(ntimes, nbl_slice, nfreqs, npols)`` otherwise.
+    flags : np.ndarray
+        Flags to write to the file. Must have the same shape as ``data``.
+    nsamples : np.ndarray
+        Number of samples to write to the file. Must have the same shape as ``data``.
+    time_index : int, optional
+        The time index to write the data to. If None, the data is written to all times.
+    """
+    if not isinstance(fl, FastUVH5Meta):
+        fl = FastUVH5Meta(fl)
+
+    # The file must be in specific format to use this specialized function.
+    if not fl.time_axis_faster_than_bls:
+        raise NotImplementedError(
+            "write_baseline_slc_to_file only works for files with shape ``(nbls, ntimes, ...)``."
+        )
+
+    if time_index is None:
+        for time_index in range(fl.Ntimes):
+            write_baseline_slc_to_file(
+                fl, baseline_slice, data[time_index], flags[time_index], nsamples[time_index], time_index
+            )
+    else:
+        slc = slice(
+            time_index + baseline_slice.start * fl.Ntimes,
+            time_index + baseline_slice.stop * fl.Ntimes,
+            fl.Ntimes
         )
+        fl.close()
+        with h5py.File(fl.path, 'a') as _fl:
```

Comment:
Why are we doing this instead of using the partial i/lo already built into hera_cal.io and into UVData? Really want to keep all file-format particulars to pyuvdata

Reply:
I had a look at the partial writing in hera_cal.io but I couldn't tell if it would do what I needed here. I would be happy to use existing infrastructure if available. 

I see that pyuvdata has a function to do this, actually. It looks _way_ more complicated than my simple function, and is probably slower, but I think you're right that we should be using it (and speeding it up if it's slow for us).

Reply:
I've updated this to use the pyuvdata functionality. It shouldn't actually be too much slower because it allows passing the `blt_inds` directly.